src/strands/session/s3_session_manager.py

```diff
@@ -0,0 +1,237 @@
+        previous_agent = self.read_agent(session_id=session_id, agent_id=agent_id)
+        if previous_agent is None:
+            raise SessionException(f"Agent {agent_id} in session {session_id} does not exist")
+
+        # Preserve creation timestamp
+        session_agent.created_at = previous_agent.created_at
+        agent_key = f"{self._get_agent_path(session_id, agent_id)}agent.json"
+        self._write_s3_object(agent_key, asdict(session_agent))
+
+    def create_message(self, session_id: str, agent_id: str, session_message: SessionMessage) -> None:
+        """Create a new message in S3."""
+        message_id = session_message.message_id
+        message_dict = asdict(session_message)
+        message_key = self._get_message_path(session_id, agent_id, message_id)
+        self._write_s3_object(message_key, message_dict)
+
+    def read_message(self, session_id: str, agent_id: str, message_id: str) -> Optional[SessionMessage]:
+        """Read message data from S3."""
+        message_key = self._get_message_path(session_id, agent_id, message_id)
+        message_data = self._read_s3_object(message_key)
+        if message_data is None:
+            return None
+        return SessionMessage.from_dict(message_data)
+
+    def update_message(self, session_id: str, agent_id: str, session_message: SessionMessage) -> None:
+        """Update message data in S3."""
+        message_id = session_message.message_id
+        previous_message = self.read_message(session_id=session_id, agent_id=agent_id, message_id=message_id)
+        if previous_message is None:
+            raise SessionException(f"Message {message_id} does not exist")
+
+        # Preserve creation timestamp
+        session_message.created_at = previous_message.created_at
+        message_key = self._get_message_path(session_id, agent_id, message_id)
+        self._write_s3_object(message_key, asdict(session_message))
+
+    def list_messages(
+        self, session_id: str, agent_id: str, limit: Optional[int] = None, offset: int = 0
+    ) -> List[SessionMessage]:
+        """List messages for an agent with pagination from S3."""
+        messages_prefix = f"{self._get_agent_path(session_id, agent_id)}messages/"
+        try:
+            paginator = self.client.get_paginator("list_objects_v2")
+            pages = paginator.paginate(Bucket=self.bucket, Prefix=messages_prefix)
+
+            # Read all message objects
+            messages: List[SessionMessage] = []
+            for page in pages:
+                if "Contents" in page:
+                    for obj in page["Contents"]:
+                        if obj["Key"].endswith(".json"):
+                            message_data = self._read_s3_object(obj["Key"])
+                            if message_data:
+                                messages.append(SessionMessage.from_dict(message_data))
+
+            # Sort by created_at timestamp (oldest first)
+            messages.sort(key=lambda x: x.created_at)
```

Comment:
Esp for S3, I think we should consider putting a timestamp in the name or something.  If we're taking in an offset + limit, it seems very inefficient to have to read all messages if we're only retrieving the last N, for instance